### ISSUE TITLE
Replace puppeteer with something more lightweight to generate PDF

### ISSUE THREAD

**Issue by ribose-jeffreylau** [riboseinc/metanorma#29]
Reason: puppeteer downloads Chrome that is nearly 300MB, whereas _e.g._ PhantomJS (and its dependencies) is only 12MB.

<s>
This file size will allow us to run within the limits of AWS Lambda.

Related links:

- http://phantomjs.org/download.html
- http://phantomjs.org/screen-capture.html
- https://aboullaite.me/phantomjs-generate-pixel-perfect-pdf-reports/
- https://coderwall.com/p/5vmo1g/use-phantomjs-to-create-pdfs-from-html

There is a [Ruby gem for phantomjs](https://rubygems.org/gems/phantomjs/versions/2.1.1.0) but it was last published in 2016, and seems to depend on capybara which we don't need.
</s>

**UPDATE**: [PhantomJS is abandonware](https://github.com/ariya/phantomjs/issues/15344).  Changing scope of this issue.

To use another headless JS implementation that meets the following requirements:

- is under active development,
- has a small enough footprint (which excludes anything that requires downloading Chrome),
  - e.g. anything over the 100MB territory is too big
  - This is to make sure it can run inside an AWS Lambda.
- is able to produce acceptable PDF outputs.

The following are nice-to-haves, but not necessary:

- exists as a Rubygem or a node.js package, or
- is easy to install from source.

**Comment by opoudjis** [comment]
brew install phantomjs: will need to update https://github.com/riboseinc/metanorma-macos-setup

Someone else will need to update https://github.com/riboseinc/metanorma-linux-setup and https://github.com/riboseinc/metanorma-windows-setup

**Comment by opoudjis** [comment]
Error: No available formula with the name "phantomjs" 
It was migrated from homebrew/core to homebrew/cask.
You can access it again by running:
  brew tap homebrew/cask
And then you can install it by running:
  brew cask install phantomjs

**Comment by ronaldtse** [comment]
@opoudjis so is the code now working with phantomjs? I can make the changes to the linux one.

**Comment by ronaldtse** [comment]
@opoudjis it may be actually better to use the Ruby gem version (forked) of phantomjs, since we can directly manage its executable and path rather than relying on the different platforms.

**Comment by opoudjis** [comment]
I did not get to this last night, will get to this tonight. I also am withholding judgement as to whether this is a good idea: didn't we reject PhantomJS because its output looked crap compared to puppeteer?

**Comment by ronaldtse** [comment]
We also need to implement some error catching in the metanorma gem so that the lack of phantomjs in the system will not cause a crash, but a warning message so that the user can remediate. HTML and Word documents should still be produced even if phantomjs isn't installed.

**Comment by opoudjis** [comment]
https://github.com/ariya/phantomjs/issues/15344 : Phantomjs is abandonware.

**Comment by opoudjis** [comment]
I've generated PDF now in metanorma using both puppeteer and phantomJS.

The phantomJS output is unacceptable. See for yourself.

[rfc6350.phantomjs.pdf](https://github.com/riboseinc/metanorma/files/2628962/rfc6350.phantomjs.pdf)
[rfc6350.puppeteer.pdf](https://github.com/riboseinc/metanorma/files/2628963/rfc6350.puppeteer.pdf)

**Comment by opoudjis** [comment]
it's possible we could make the choice between PhantomJS and puppeteer a config option somehow. But the default has to be puppeteer.

**Comment by opoudjis** [comment]
FWIW, the phantomjs gem doesn't actually require capybara; that's a development dependency, and is only triggered if you invoke poltergeist.

**Comment by ribose-jeffreylau** [comment]
Hi @opoudjis , thanks for the awesome findings.

Given the abandonware state of phantomjs, perhaps we should change the scope of this issue: to use another headless JS implementation that meets the following requirements:

- is under active development,
- has a small enough footprint (which excludes anything that requires downloading Chrome),
  - e.g. anything over the 100MB territory is too big
  - This is to make sure it can run inside an AWS Lambda.
- is able to produce acceptable PDF outputs.

The following are nice-to-haves, but not necessary:

- exists as a Rubygem or a node.js package, or
- is easy to install from source.

FWIW we initially named phantomjs just because the example PDF outputs found online (those links in the first post) looked good 😛

**Comment by ronaldtse** [comment]
Darn. In any case, apparently puppeteer could be made to run within Lambda as an external download.

https://github.com/sambaiz/puppeteer-lambda-starter-kit

**Comment by ronaldtse** [comment]
Sorry @opoudjis so let's get rid of phantomjs, not an option anymore since it's discontinued.

**Comment by opoudjis** [comment]
I'm going through the list in https://github.com/dhamaniasad/HeadlessBrowsers, but there is not a lot there: almost all of these are Javascript testbeds, not actual browsers, let alone PDF renderers.

WeasyPrint had promise, but (a) depends on icu4c for processing Unicode, which is 63MB on Mac; (b) does not understand CSS enough to suppress the sidebar; (c) does not seem able to process Webfonts; (d) does not understand enough Javascript to process Asciimath.

[rice.weasyprint.pdf](https://github.com/riboseinc/metanorma/files/2635824/rice.weasyprint.pdf)

**Comment by opoudjis** [comment]
The wkhtmltopdf precompiled binary is 50 MB. Its PDF is almost acceptable, though it mangles the tables: it understands enough Javascript to do Asciimath correctly, though not quite enough to deal with dynamic table of contents (but we're changing that anyway). The table mangling could be addressed with a CSS customisation for `@media print`. 

So it's *an* option, but I don't see it being a space saving over the Lambda version of Chromium that @ronaldtse found.

[rice.wkhtmltopdf.pdf](https://github.com/riboseinc/metanorma/files/2635831/rice.wkhtmltopdf.pdf)

**Comment by opoudjis** [comment]
ghost.py is probably abandonware: https://github.com/jeanphix/Ghost.py/issues/336 https://github.com/jeanphix/Ghost.py/issues/340. It depends on QT as a rendering engine, which is a 300MB download. Couldn't get it to install on OSX, the python installer couldn't find cmake even when I installed it (and the installer is python2, so it's old). I'm not spending more time trying to get it to work.
